### ISSUE TITLE
GPU-based Particles running slow on macOS due to transform feedback being implemented on the CPU

### ISSUE THREAD

**Issue by djrain** [godotengine/godot#54052]
### Godot version

3.3.2 stable, 3.4 RC1

### System information

macOS Big Sur
GLES3

### Issue description

I was working in my main project when I noticed some frame drops when I added more than about 20 Particles2D nodes at the same time. This seemed unreasonably slow, so I made a test scene in a new project that instances 500 basic particle systems with only 1 particle each. On my 2020 M1 Mac Mini this scene runs consistently at a ridiculous 10 FPS. I tested this in 3.3.2 stable as well as 3.4 RC1, same results on both.

For comparison, running this same scene in 3.3.2 on my 2015 MacBook Pro gets a solid 30 FPS, which seems reasonable for an older laptop with integrated graphics. Also, some fellow devs testing the same code on PC and Linux had no issues. So it seems this may be an issue specific to M1 machines.

### Steps to reproduce

run the test project, presumably on an M1 Mac

### Minimal reproduction project

[particle performance test.zip](https://github.com/godotengine/godot/files/7385056/particle.performance.test.zip)

**Comment by floppyhammer** [comment]
Got similar result on the same machine.

![20211021-100014](https://user-images.githubusercontent.com/28705694/138198220-a538f64f-438b-421d-add8-9a804fc84971.png)

**Comment by Calinou** [comment]
> Unfortunately this may mean that CPUParticles are the only viable option for particles on M1 macs (in 3.x that is, in 4.0 the Vulkan/Metal renderer may work much better)

Don't GPUParticles in `master` use transform feedback too?

**Comment by Richard74Huang** [comment]
Having same issue here on M1 with Godot 3.4 stable. Also tried to export as release but still laggy.

**Comment by clayjohn** [comment]
I am fairly certain this is caused by Apples poor support of the OpenGL standard. Specifically, I don't think Apple devices support using transform feedback on the GPU, so the drivers emulate it by passing data back to the CPU. If that is the case then the solution is to run OpenGL over Metal on Apple devices. There is a draft PR already: https://github.com/godotengine/godot/pull/50253

**Comment by Richard74Huang** [comment]
Apple joins Blender Dev team recently. If they really want to promote their apple silicon, helping  game developers like Godot engine is crucial too. 

Anyway, I'll use CPUParticle2D instead on macOS for now. Thanks :)

**Comment by Calinou** [comment]
I wonder if we should expose the "Convert to CPUParticles" option to be used at run-time, and run it automatically on all GPU-based particle nodes by default on macOS.

**Comment by akien-mga** [comment]
Closed by #55268.

**Comment by djrain** [comment]
I'm not quite convinced that we've found the real issue here.

Even on my M1 Mac, as I mentioned previously, a single Particles2D is still much more capable than a single CPUParticles2D. It can do 1 million particles at almost 60 fps. It's only when I instance _numerous_ Particles2D nodes that the performance degrades. And the number of nodes has much more of an impact on the frame rate than the total number of particles. The fact that only 100 Particles2D nodes, emitting _one_ particle each, is taking the fps down to 39... just seems very odd. Is there an explanation?

**Comment by clayjohn** [comment]
Yes, same explanation as above. TransformFeedback is implemented on the CPU in apple's OpenGL driver, so  to update each Particles2D node the entire GPU process stalls while the particle data is passed from GPU to CPU and back to GPU. 

It is more efficient to transfer a million particles once than it is send a single particle a thousand times. That is why things like batching can be effective. It is more efficient to send data to and from the GPU in big batches than it is to send a thousand tiny commands.

**Comment by djrain** [comment]
I see, so sending from GPU to CPU and back is just that slow... and CPUParticles is much faster because information is only being sent one way (to GPU)?

If batching is a potential solution, how can I make sure that happens? I'm setting all the instances to the same ParticlesMaterial, no texture/material and tiny visibility rect (is this used for overlap test?), but it isn't batching:

items
	joined_item 1 refs, 
			batch D 0-1 PA 
	joined_item 1 refs, 
			batch D 0-1 PA 
	joined_item 1 refs, 
			batch D 0-1 PA

**Comment by clayjohn** [comment]
> I see, so sending from GPU to CPU and back is just that slow... and CPUParticles is much faster because information is only being sent one way (to GPU)?

That is my understanding, yes. I don't know much about how Transform Feedback is implemented on Apple devices (other than that it is in software) so there could be some other things that the driver does that explain the poor performance.

> If batching is a potential solution, how can I make sure that happens? I'm setting all the instances to the same ParticlesMaterial, no texture/material and tiny visibility rect (is this used for overlap test?), but it isn't batching:

Particles are not batched. If you only have a single particle in each Particles2D then it will be much more efficient to use a Sprite instead. For most hardware, Particles only become worthwhile once you start having hundreds to thousands of instances within a Particles2D (hundreds on lower end hardware, thousands on higher end hardware). 

My guess is that 1000 Sprites will outperform 1000 Particles2D.

You have three alternatives to try out:
1. Use fewer Particles2D nodes and increase the number of particles in each,
2. Use CPUParticles where possible (also with a high number of particles in each),
3. Use regular Sprites (if sharing a material and texture these should batch automatically).

I can't say for sure what will perform the best for you. But it is worth trying a few different approaches to see what works best for your workflow and performance goals.

**Comment by djrain** [comment]
@clayjohn Thanks, good to know! As of now I'm using VisualServer for particles as well as several thousand stars, and it's running like a dream :)

**Comment by lekoder** [comment]
@akien-mga I'm not sure this is a good solution to this issue. Please correct me if I'm wrong, but you choose to solve "particles are working slow on M1" by having "if the developer happens to work on M1, warn them about that".

It doesn't really fix anything for multi-platform games, which are typically not developed on the target platform. Adding the notice when you export the game would at least notify the developer of the problem, but you cannot reasonably expect developer to re-make entire particle system in their game, which might be hundreds of nodes across hundreds of scenes, just for sake of targeting additional platform - which will additionally decrease the performance for the unaffected platforms.

A proper solution would be to offer conversion from Particles2D to CPUParticles2D either at runtime or at export time.

**Comment by akien-mga** [comment]
Well there doesn't seem to be a good solution to that issue indeed, at least until a macOS focused rendering contributor decides to look into what workaround could be implemented for that platform.

A conversion to CPUParticles2D sounds interesting but AFAIK the features don't map 1:1 so I'm not sure how well this would work. Anyways reopening for further discussion.

**Comment by lekoder** [comment]
I think at the very least it should present a warning when you target a OSX export regardless of the platform you are developing on. This would make the developer aware of the problem and they can work around it.

I intend to solve that by making a container to sibling `Particles2D` and `CPUParticles2D` placeholders, with a unified API exposed by the container, but probably the ideal solution would be to get a full compatibility of `CPUParticles2D` and `Particles2D` ; that would allow both the ability convert them on export or perhaps a run-time option to switch between them.

**Comment by Calinou** [comment]
I think we should add a project setting that automatically converts GPUParticles3D to CPUParticles3D at run-time on macOS. It's better to have broken particles than unplayable performance when the project runs on macOS.

**Comment by GeraldineSullivan** [comment]
This is happening to me today on M2 macbook pro running Sonoma

**Comment by Calinou** [comment]
> This is happening to me today on M2 macbook pro running Sonoma

Which Godot version and rendering method are you using?

**Comment by GeraldineSullivan** [comment]
> > This is happening to me today on M2 macbook pro running Sonoma
> 
> Which Godot version and rendering method are you using?

I am using 4.1.2

I fixed it by using the renderer settings from the test file in OP. They were all set to gl_compatibility by default.

**Comment by findthebug** [comment]
Macbook Air M3 . 24 GB Ram - Same issue.

As soon as i add a basic CPUParticles2D, framerate drops to 10 fps after some time.  Using a lot of physics in the scene. Only solution is Forward+.
